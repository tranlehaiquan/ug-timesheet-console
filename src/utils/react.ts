import * as React from 'react'

type IdentityFunction = <T>(fn: T) => T

export const typedMemo: IdentityFunction = React.memo

export const getDisplayName = <T extends any = any>(WrappedComponent: React.ComponentType<T>): string => {
  return WrappedComponent.displayName || WrappedComponent.name || 'Component'
}
