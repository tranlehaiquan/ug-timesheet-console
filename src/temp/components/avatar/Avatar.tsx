import * as React from "react";
import classnames from "classnames";
import { omit } from "lodash";

import { Tooltip } from "../popups/tooltip/Tooltip";

type AvatarSizes = "tiny" | "small" | "medium" | "large";
export interface IAvatar extends React.HTMLAttributes<HTMLDivElement> {
  /** This will be used for the 'letter' display and be the
   * default tooltip text if showTooltip is true
   * */
  name: string;
  /**
   * The url to use to fetch the avatar image
   */
  imageUrl?: string;
  /**
   * 'medium' is the default
   */
  size?: AvatarSizes;
  /**
   * Wrap the avatar in a tooltip
   */
  showTooltip?: boolean;
  /**
   * If you want your tooltip text to say something other
   * than the 'name' prop, you can supply the text to this prop
   */
  tooltipText?: string;
}

/**
 * Simple component for displaying avatars
 * @requires Tooltip
 */
export class Avatar extends React.PureComponent<IAvatar, {}> {
  sizeClassMap: { [key in AvatarSizes]: string } = {
    tiny: "sked-avatar--tiny",
    small: "sked-avatar--small",
    medium: "",
    large: "sked-avatar--large",
  };

  getInitials(name: string = "") {
    if (name.length > 1) {
      const [first, last] = name.trim().split(" ");
      return (
        first && last
          ? `${first.charAt(0)}${last.charAt(0)}`
          : first.substring(0, 2)
      ).toUpperCase();
    }
    return name.toUpperCase();
  }

  wrapInTooltip(avatar: JSX.Element) {
    return (
      <Tooltip
        content={this.props.tooltipText || this.props.name}
        position={"top"}
        delayShow={250}
      >
        {avatar}
      </Tooltip>
    );
  }

  renderAvatar() {
    const { name, size, className, imageUrl, ...rest } = this.props;
    const otherProps = omit(rest, ["showTooltip", "tooltipText"]);
    return (
      <div
        className={classnames(
          "sked-avatar",
          className,
          this.sizeClassMap[size]
        )}
        {...otherProps}
      >
        {imageUrl ? (
          <img className="sked-avatar__image" src={imageUrl} />
        ) : (
          this.getInitials(name)
        )}
      </div>
    );
  }

  render() {
    return this.props.showTooltip
      ? this.wrapInTooltip(this.renderAvatar())
      : this.renderAvatar();
  }
}
