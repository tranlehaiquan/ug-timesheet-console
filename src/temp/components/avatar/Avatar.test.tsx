import * as React from 'react'
import { Avatar } from './Avatar'
import { shallow, render } from 'enzyme'

describe('Avatar', () => {
  it('renders correctly', () => {
    const avatar = render(<Avatar name="Elliot Alderson" />)

    expect(avatar).toMatchSnapshot()
  })

  it('should use first initials of passed in name', () => {
    const avatar = shallow(<Avatar name="Angela moss" data-sk-name="test" />)

    expect(avatar.find('[data-sk-name="test"]').text()).toEqual('AM')
  })

  it('should use the first 2 letters of a name if there is no last name', () => {
    const avatar = shallow(<Avatar name=" Angela" data-sk-name="test" />)

    expect(avatar.find('[data-sk-name="test"]').text()).toEqual('AN')
  })

  it('should not apply a tooltip unless required', () => {
    const avatar = shallow(<Avatar name="Mr Robot" />)

    expect(avatar.find('Tooltip').length).toEqual(0)
  })

  it('should apply a tooltip if required', () => {
    const avatar = shallow(<Avatar name="Darlene" showTooltip />)

    expect(avatar.find('Tooltip').length).toEqual(1)
    expect(avatar.find('Tooltip').prop('content')).toEqual('Darlene')
  })

  it('should use supplied tooltip text', () => {
    const avatar = shallow(<Avatar name="Darlene" showTooltip tooltipText="Hello" />)

    expect(avatar.find('Tooltip').prop('content')).toEqual('Hello')
  })

  it('displays an image if url is supplied', () => {
    const avatar = shallow(<Avatar name="Darlene" imageUrl="./test.png" />)

    expect(avatar.find('img').prop('src')).toEqual('./test.png')
  })

  it('applies image class if using image', () => {
    const avatar = shallow(<Avatar name="Darlene" imageUrl="./test.png" />)

    expect(avatar.find('img').hasClass('sked-avatar__image')).toBe(true)
  })

  it('applies correct sizing classes', () => {
    const avatar = shallow(<Avatar name="Darlene" data-sk-name="test" />)

    expect(avatar.find('[data-sk-name="test"]').hasClass('sked-avatar')).toBe(true)

    avatar.setProps({ size: 'tiny' })
    expect(avatar.find('[data-sk-name="test"]').hasClass('sked-avatar--tiny')).toBe(true)

    avatar.setProps({ size: 'small' })
    expect(avatar.find('[data-sk-name="test"]').hasClass('sked-avatar--small')).toBe(true)

    avatar.setProps({ size: 'large' })
    expect(avatar.find('[data-sk-name="test"]').hasClass('sked-avatar--large')).toBe(true)
  })
})
