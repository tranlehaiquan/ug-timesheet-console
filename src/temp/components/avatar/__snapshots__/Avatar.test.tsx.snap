// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar renders correctly 1`] = `
<div
  class="sked-avatar"
>
  EA
</div>
`;
